tcmur_cmd_handler.c

```diff
@@ -2308,6 +2318,312 @@ static int handle_inquiry(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+		.opcode = WRITE_VERIFY_16,
+		.cdb_length = 16,
+		.cdb_usage = {WRITE_VERIFY, 0xf2, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 
+		              0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, CONTROL_BYTE},
+	};
+
+	static cdb_info opcode_format_unit = {
+		.opcode = FORMAT_UNIT,
+		.cdb_length = 6,
+		.cdb_usage = {FORMAT_UNIT, 0x00, 0xff, 0xff, 0xff, CONTROL_BYTE},
+	};
+
+	switch (op) {
+	case INQUIRY:
+		buf = opcode_inquiry.cdb_usage;
+		break;
+	case TEST_UNIT_READY:
+		buf = opcode_test_unit_ready.cdb_usage;
+		break;
+	case SERVICE_ACTION_IN_16:
+		buf = opcode_service_action_in_16.cdb_usage;
+		break;
+	case READ_CAPACITY:
+		buf = opcode_read_capacity.cdb_usage;
+		break;
+	case MODE_SENSE:
+		buf = opcode_mode_sense.cdb_usage;
+		break;
+	case MODE_SENSE_10:
+		buf = opcode_mode_sense_10.cdb_usage;
+		break;
+	case START_STOP:
+		buf = opcode_start_stop.cdb_usage;
+		break;
+	case MODE_SELECT:
+		buf = opcode_mode_select.cdb_usage;
+		break;
+	case MODE_SELECT_10:
+		buf = opcode_mode_select_10.cdb_usage;
+		break;
+	case RECEIVE_COPY_RESULTS:
+		buf = opcode_receive_copy_results.cdb_usage;
+		break;
+	case MAINTENANCE_IN:
+		switch (sa) {
+		case MI_REPORT_TARGET_PGS:
+			buf = opcode_mi_report_target_pgs.cdb_usage;
+			break;
+		case MI_REPORT_SUPPORTED_OPERATION_CODES: 
+			buf = opcode_mi_supported_operation_codes.cdb_usage;
+			break;
+		default:
+			break;
+		}
+		break;
+	case UNMAP:
+		buf = opcode_unmap.cdb_usage;
+		break;
+	case EXTENDED_COPY:
+		buf = opcode_extended_copy.cdb_usage;
+		break;
+	case READ_16:
+		buf = opcode_read_16.cdb_usage;
+		break;
+	case WRITE_16:
+		buf = opcode_write_16.cdb_usage;
+		break;
+	case WRITE_SAME_16:
+		buf = opcode_write_same_16.cdb_usage;
+		break;
+	case WRITE_VERIFY_16:
+		buf = opcode_write_verify_16.cdb_usage;
+		break;
+	case FORMAT_UNIT:
+		buf = opcode_format_unit.cdb_usage;
+		break;
+	default:
+		break;
+	}
+
+	return buf;
+}
+
+int is_support_opcode(int opcode) 
+{
+	switch (opcode) {
+	case INQUIRY:
+	case TEST_UNIT_READY:
+	case SERVICE_ACTION_IN_16:
+	case READ_CAPACITY:
+	case MODE_SENSE:
+	case MODE_SENSE_10:
+	case START_STOP:
+	case MODE_SELECT:
+	case MODE_SELECT_10:
+	case RECEIVE_COPY_RESULTS:
+	case MAINTENANCE_IN:
+	case UNMAP:
+	case EXTENDED_COPY:
+	case READ_16:
+	case WRITE_16:
+	case WRITE_SAME_16:
+	case WRITE_VERIFY_16:
+	case FORMAT_UNIT:
+		return 1;
+	default:
+		return 0;
+	} 
+}
+
+int report_opcode_one(struct tcmulib_cmd *cmd, int rctd, uint8_t opcode, 
+                      uint16_t service_action, int is_service_action)
+{
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	int cdb_len = get_cdb_length(opcode);
+	uint8_t buf[256], *ptr, *usage_data;
+	
+	if ((!is_support_opcode(opcode)) || (is_service_action && !service_action) ||
+           (!is_service_action && service_action)) 
+		return SAM_STAT_CHECK_CONDITION;
+
+	memset(buf, 0, sizeof(buf));
+	ptr = &buf[0];
+	ptr++;
+	*ptr++ = rctd ? 0x83 : 0x03;
```

Comment:
Could you add a comment about what this bit is here and below. Just maybe the same of
the bit in plain english like it is in the spec. No one remembers what
these are off the top of their head so it makes reading the code a little
easier. Also if there is a mistake we can at least figure out what we were
trying to do :)  Do that below too.